### PR TITLE
MQTT: Use original topic for subscription and removal

### LINKED ISSUE #4739: MqttBrokerConnection.removeConsumer() is not removing the consumer
I'm facing a problem while using org.eclipse.smarthome.io.transport.mqtt.

If I use addConsumer with a subscriber using the topic "+/+/+/+/+/+" and later on try to remove the same subscriber, the subscriber is not really removed from the list of consumers in MqttBrokerConnection.

In addConsumer() there is a conversion of the topic:
`String topic = StringUtils.replace(StringUtils.replace(subscriber.getTopic(), "+", "[^/]*"), "#", ".*");`

And later on `topic` is used, while in removeConsumer `subscriber.getTopic()` is used.

If the above conversion is added to removeConsumer() it seems to fix the problem.

### PR DESCRIPTION
**Pull request by triller-telekom**
Fixes #4739

Signed-off-by: Stefan Triller <stefan.triller@telekom.de>

### CODE REVIEW

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java [hunk 1]
```diff
@@ -360,7 +360,7 @@ public boolean addConsumer(MqttMessageSubscriber subscriber) throws MqttExceptio
                 subscriberList = new ArrayList<>();
             }
             subscriberList.add(subscriber);
-            consumers.put(topic, subscriberList);
+            consumers.put(subscriber.getTopic(), subscriberList);
```
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java]: The current state of the patch does not make sense to me.
Five lines above a look up is done to get the `subscriberList` and a new one is created if there is no one.
After that the subscriber is added to that list and put to the consumers using the key `subscriber.getTopic()`.
The lookup above it done using still the key `topic`.

So, if the original topic should be used, it needs to be fixed on all places.

I would also prefer to not use put if we already receive the list. We can add the element to the list without a further put operation.

Comment (maggu2810) [reply bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java]: Another question is why we need to replace parts of the topic at all in this method if the key consists of the original topic...

Comment (triller-telekom) [reply bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java]: > So, if the original topic should be used, it needs to be fixed on all places.

I have done that now

> I would also prefer to not use put if we already receive the list. We can add the element to the list without a further put operation.

Done.

> Another question is why we need to replace parts of the topic at all in this method if the key consists of the original topic...

That's a very good question, the comment in code says something about that it is necessary for comparing it with a regex. But I think this PR is just for making the current behavior consistent. The topic replacement should be discussed in a new issue.

**Comment by maggu2810** [comment]
>> Another question is why we need to replace parts of the topic at all in this method if the key consists of the original topic...

> That's a very good question, the comment in code says something about that it is necessary for comparing it with a regex. But I think this PR is just for making the current behavior consistent. The topic replacement should be discussed in a new issue.

Okay for me. Will you create a new issue and add `@...` for the one that introduced the substring replacement?

**Comment by triller-telekom** [comment]
I have opened #4769 regarding the substring replacement.

**Comment by maggu2810** [comment]
@triller-telekom Thank you

**Pull request merged**